subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/vcs/VcsDependencyResolver.java

```diff
@@ -0,0 +1,207 @@
+        this.serviceRegistry = serviceRegistry;
+        this.vcsMappingsInternal = vcsMappingsInternal;
+        this.vcsMappingFactory = vcsMappingFactory;
+        this.versionControlSystemFactory = versionControlSystemFactory;
+        // TODO: We need to have a "cache" for repositories in the VCS implementation and a working dir "cache" for included builds
+        // This path shouldn't be hardcoded here
+        this.cacheDir = new File("build");
+    }
+
+    @Override
+    public DependencyToComponentIdResolver getComponentIdResolver() {
+        return this;
+    }
+
+    @Override
+    public ComponentMetaDataResolver getComponentResolver() {
+        return this;
+    }
+
+    @Override
+    public OriginArtifactSelector getArtifactSelector() {
+        return this;
+    }
+
+    @Override
+    public ArtifactResolver getArtifactResolver() {
+        return this;
+    }
+
+    @Override
+    public void resolve(DependencyMetadata dependency, BuildableComponentIdResolveResult result) {
+        if (dependency.getSelector() instanceof SourceComponentSelector) {
+            result.failed(new ModuleVersionResolveException(dependency.getSelector(), "not implemented"));
+            // TODO: Handle selectors that have already been created by some other means
+//            ProjectComponentSelector selector = (ProjectComponentSelector) dependency.getSelector();
+//            ProjectComponentIdentifier project = componentIdentifierFactory.createProjectComponentIdentifier(selector);
+//            LocalComponentMetadata componentMetaData = localComponentRegistry.getComponent(project);
+//            if (componentMetaData == null) {
+//                result.failed(new ModuleVersionResolveException(selector, project + " not found."));
+//            } else {
+//                result.resolved(componentMetaData);
+//            }
+        } else {
+            VcsMappingInternal vcsMappingInternal = getVcsMapping(dependency);
+            if (vcsMappingInternal != null) {
+                // TODO: Hardcoded version of "master"
+                // This should be based on something from the repository.
+                // TODO: Need to clone the repository now and get the version information?
+                // TODO: Need failure handling, e.g., cannot clone repository
+                vcsMappingsInternal.getVcsMappingRule().execute(vcsMappingInternal);
+                if (vcsMappingInternal.isUpdated()) {
+                    String version = "master";
+                    String projectPath = ":"; // assume it's from the root
+                    String buildName = vcsMappingInternal.getOldRequested().getName();
+                    result.resolved(DefaultSourceComponentIdentifier.newIdentifier(vcsMappingInternal.getRepository(), projectPath, new DefaultBuildIdentifier(buildName)),
```

Comment:
For 2a, shouldn't we prefer to use binary dependencies we've found from some repository at a given version rather than building those same artifacts from source? I'm thinking that we should be consulting the `Resolver` implementations which search binary repositories earlier in the chain than this `VCSDependencyResolver`. But, if we did that, I'm worried that a version selector like `latest.integration` might be satisfied by the artifact repository `Resolver` before the `VCSDependencyResolver` gets its chance to resolve it. Maybe we need a new way to designate the version. Maybe `branch.<branchname>` or something like that?

Reply:
We can make either source wins or binary wins work.

For dynamic selectors like `latest.integration`, Gradle will ask each resolver and then select the highest match it finds across all of them. The VCS resolver can communicate in the metadata that the thing built from master is of a version that is higher than all other versions (and so should be selected in preference to older things). Later we can make this better to deal with branches and binaries in a repo that are already built from master and things like that.